File: 82-entity-value-object/EntityValueDemo.cpp
/*
 * Ngôn ngữ: C++
 * Công dụng: Bản C++ của cùng bài học — hai phép thử để quyết định entity hay value
 * object, và hai con bug thật: định danh do CSDL cấp làm hai khách hàng gộp làm một,
 * và value object khả biến bị chia sẻ làm sai địa chỉ giao hàng.
 * Tại sao cần học: C++ đảo ngược bài toán so với Java và Python. Ở đó, mặc định là
 * chia sẻ tham chiếu nên VALUE OBJECT là thứ phải cố gắng làm đúng. Ở C++, mặc định
 * là ngữ nghĩa giá trị — copy một struct là copy thật — nên value object gần như cho
 * không, còn ENTITY mới là thứ nguy hiểm: một dòng `auto ban_sao = *kho;` sinh ra hai
 * object CÙNG ĐỊNH DANH nhưng khác trạng thái, một mâu thuẫn nghiệp vụ không thể tồn
 * tại ngoài đời. Bù lại, C++ cho công cụ dập tắt nó ở mức trình biên dịch: `= delete`.
 */
#include <iostream>
#include <memory>
#include <set>
#include <stdexcept>
#include <string>
#include <vector>
#include <cstdlib>

// =====================================================================
// VALUE OBJECT — trong C++, day la mac dinh cua ngon ngu
// =====================================================================
class Tien {
public:
    Tien(long soTien, std::string tienTe) : soTien_(soTien), tienTe_(std::move(tienTe)) {
        if (soTien_ < 0) throw std::invalid_argument("so tien khong duoc am");
        if (tienTe_.size() != 3) throw std::invalid_argument("ma tien te phai dung 3 ky tu");
    }

    long soTien() const { return soTien_; }
    const std::string& tienTe() const { return tienTe_; }

    // Value object mang LUAT cua chinh no, khong chi mang du lieu.
    Tien cong(const Tien& khac) const {
        if (tienTe_ != khac.tienTe_)
            throw std::invalid_argument("khong cong duoc hai loai tien te");
        return Tien(soTien_ + khac.soTien_, tienTe_);   // TRA VE CAI MOI
    }

    // So sanh theo TOAN BO gia tri — dinh nghia cua value object.
    bool operator==(const Tien& k) const { return soTien_ == k.soTien_ && tienTe_ == k.tienTe_; }
    bool operator<(const Tien& k) const {
        return tienTe_ != k.tienTe_ ? tienTe_ < k.tienTe_ : soTien_ < k.soTien_;
    }

private:
    // Khong `const` — xem phan 7 de biet vi sao. Bat bien duoc bao dam bang viec
    // KHONG CO SETTER, khong phai bang `const` tren field.
    long soTien_;
    std::string tienTe_;
};

class DiaChi {
public:
    DiaChi(std::string duong, std::string phuong, std::string tinh)
        : duong_(std::move(duong)), phuong_(std::move(phuong)), tinh_(std::move(tinh)) {
        if (duong_.empty()) throw std::invalid_argument("duong khong duoc rong");
    }
    const std::string& duong() const { return duong_; }

    // "Doi" mot value object = tao cai moi. Khong co setter, khong the co.
    DiaChi voiDuong(const std::string& moi) const { return DiaChi(moi, phuong_, tinh_); }

    bool operator==(const DiaChi& d) const {
        return duong_ == d.duong_ && phuong_ == d.phuong_ && tinh_ == d.tinh_;
    }

private:
    std::string duong_, phuong_, tinh_;
};

// =====================================================================
// ENTITY — va o C++, day moi la thu phai canh giu
// =====================================================================
class DiemGiao {
public:
    DiemGiao(std::string ma, DiaChi diaChi, std::string nguoiPhuTrach)
        : ma_(std::move(ma)), diaChi_(std::move(diaChi)), nguoiPhuTrach_(std::move(nguoiPhuTrach)) {
        if (ma_.empty()) throw std::invalid_argument("diem giao phai co ma ngay luc tao");
    }

    // DAY LA DIEU CHI C++ LAM DUOC: cam sao chep entity o muc trinh bien dich.
    // Voi hai dong nay, `auto sao = *kho;` la LOI BIEN DICH:
    //     error: use of deleted function 'DiemGiao::DiemGiao(const DiemGiao&)'
    DiemGiao(const DiemGiao&) = delete;
    DiemGiao& operator=(const DiemGiao&) = delete;
    // Di chuyen thi VAN cho phep: chuyen quyen so huu khong sinh ra ban sao dinh danh.
    DiemGiao(DiemGiao&&) = default;
    DiemGiao& operator=(DiemGiao&&) = default;

    const std::string& ma() const { return ma_; }
    const DiaChi& diaChi() const { return diaChi_; }   // an toan: DiaChi bat bien
    void doiDiaChi(DiaChi moi) { diaChi_ = std::move(moi); }
    void doiNguoiPhuTrach(std::string moi) { nguoiPhuTrach_ = std::move(moi); }

    // Entity so sanh CHI theo dinh danh. Khong field nao khac duoc xuat hien o day.
    bool operator==(const DiemGiao& d) const { return ma_ == d.ma_; }
    bool operator<(const DiemGiao& d) const { return ma_ < d.ma_; }

private:
    const std::string ma_;   // dinh danh: bat bien tron doi
    DiaChi diaChi_;          // thuoc tinh: doi thoai mai
    std::string nguoiPhuTrach_;
};

// =====================================================================
// BAN SAI 1 — entity CHO PHEP sao chep, va dinh danh bi nhan ban
// =====================================================================
struct DiemGiaoSai {
    std::string ma;
    std::string nguoiPhuTrach;
    bool operator==(const DiemGiaoSai& d) const { return ma == d.ma; }
};

// =====================================================================
// BAN SAI 2 — entity lay dinh danh tu CSDL
// =====================================================================
struct KhachHangSai {
    long id = 0;              // 0 = "chua luu". Quy uoc pho bien nhat, va sai nhat.
    std::string ten;
    bool operator<(const KhachHangSai& k) const { return id < k.id; }
};

// BAN DUNG — dinh danh sinh trong mien, co ngay tu luc tao
class KhachHang {
public:
    explicit KhachHang(std::string ten) : ma_("KH-" + std::to_string(++dem_)), ten_(std::move(ten)) {}
    const std::string& ma() const { return ma_; }
    bool operator<(const KhachHang& k) const { return ma_ < k.ma_; }
private:
    static long dem_;
    std::string ma_, ten_;
};
long KhachHang::dem_ = 0;

// =====================================================================
// BAN SAI 3 — "value object" nhung co setter
// =====================================================================
struct DiaChiSai {
    std::string duong;
    void setDuong(std::string d) { duong = std::move(d); }
};

// ---- Self-check ----
static void check(bool ok, const std::string& msg) {
    if (!ok) {
        std::cerr << "FAIL: " << msg << "\n";
        std::exit(1);
    }
}

int main() {
    // ---- 1. HAI PHEP THU de quyet dinh entity hay value object ----
    //
    // Phep thu A — "doi HET thuoc tinh, con la cung mot thu khong?"
    DiemGiao kho("DG-01", DiaChi("12 Le Loi", "Ben Nghe", "TP.HCM"), "anh Nam");
    const std::string maGoc = kho.ma();
    kho.doiDiaChi(DiaChi("45 Nguyen Hue", "Ben Nghe", "TP.HCM"));
    kho.doiNguoiPhuTrach("chi Lan");
    check(kho.ma() == maGoc, "doi het thuoc tinh, van la cung diem giao -> ENTITY");

    // Phep thu B — "hai cai giong het nhau, thay cho nhau duoc khong?"
    Tien a(50000, "VND"), b(50000, "VND");
    check(a == b, "hai to 50.000d thay cho nhau duoc -> VALUE OBJECT");
    check(&a != &b, "van la hai object khac nhau trong bo nho — va dieu do KHONG quan trong");

    // ---- 2. CUNG MOT KHAI NIEM, HAI VAI TRO — tuy NGU CANH ----
    DiaChi dcDon("12 Le Loi", "Ben Nghe", "TP.HCM");
    DiaChi dcKhac("12 Le Loi", "Ben Nghe", "TP.HCM");
    check(dcDon == dcKhac, "trong don hang: hai dia chi giong nhau LA MOT");

    DiemGiao khoA("DG-01", dcDon, "anh Nam");
    DiemGiao khoB("DG-02", dcKhac, "chi Lan");
    check(!(khoA == khoB), "trong van chuyen: cung dia chi van la HAI diem giao");
    check(khoA.diaChi() == khoB.diaChi(), "du thuoc tinh dia chi cua chung bang nhau");
    // Cau hoi "cai nay la entity hay value object" KHONG co cau tra loi chung. No phu
    // thuoc vao viec nghiep vu co can theo doi CAI CU THE NAY qua thoi gian hay khong.

    // ---- 3. DIEU CHI C++ MOI CO: SAO CHEP ENTITY LA MAU THUAN NGHIEP VU ----
    // Voi ban SAI (struct thuong), sao chep bien dich ngon lanh:
    DiemGiaoSai saiGoc{"DG-99", "anh Nam"};
    DiemGiaoSai saiSao = saiGoc;               // mot dau `=` sinh ra mot the gioi song song
    saiSao.nguoiPhuTrach = "chi Lan";
    check(saiGoc == saiSao, "hai object nay LA CUNG MOT diem giao (cung ma)...");
    check(saiGoc.nguoiPhuTrach != saiSao.nguoiPhuTrach, "...nhung trang thai KHAC NHAU");
    // Doc lai hai dong tren: cung mot cai kho, dong thoi do anh Nam va chi Lan phu
    // trach. Ngoai doi khong ton tai. Trong bo nho thi ton tai — va tu day tro di, moi
    // thay doi tren mot ban se khong den ban kia. Du lieu bat dau troi ra khoi nhau.

    // Voi ban DUNG, chinh dong do KHONG BIEN DICH DUOC:
    //     DiemGiao ban_sao = kho;
    //     error: use of deleted function 'DiemGiao::DiemGiao(const DiemGiao&)'
    //
    // Nghia la o C++, "khong duoc nhan ban entity" khong phai mot quy uoc trong tai
    // lieu — no la mot luat trinh bien dich thuc thi. Java va Python khong lam duoc dieu
    // nay; o do ban chi co the... nho dung goi `clone()`.
    //
    // He qua: entity phai luu qua con tro / tham chieu, khong luu theo gia tri.
    std::vector<std::unique_ptr<DiemGiao>> cacKho;
    cacKho.push_back(std::make_unique<DiemGiao>("DG-10", dcDon, "anh Nam"));
    cacKho.push_back(std::make_unique<DiemGiao>("DG-11", dcKhac, "chi Lan"));
    check(cacKho.size() == 2, "entity luu bang unique_ptr — mot object, mot chu so huu");
    // Nguoc lai, value object luu THANG theo gia tri, va do la cach dung:
    std::vector<Tien> vi{Tien(50000, "VND"), Tien(20000, "VND")};
    check(vi.size() == 2, "value object luu theo gia tri, sao chep thoai mai");

    // ---- 4. CON BUG: dinh danh do CSDL cap -> hai khach gop lam mot ----
    std::set<KhachHangSai> gioSai{KhachHangSai{0, "Nguyen Van A"}, KhachHangSai{0, "Tran Thi B"}};
    check(gioSai.size() == 1, "HAI khach hang khac nhau, set chi giu MOT");
    // Mat mot khach hang. Khong ngoai le, khong log. Bug nay chi xuat hien khi xu ly
    // theo lo (nhap file CSV, tao hang loat) — nghia la no qua duoc moi test thu cong.

    std::set<KhachHang> gioDung{KhachHang("Nguyen Van A"), KhachHang("Tran Thi B")};
    check(gioDung.size() == 2, "dinh danh sinh trong mien -> hai khach, giu dung hai");
    // He qua thuc te: test khong can CSDL, va co the gui entity qua hang doi truoc khi
    // luu — thu ma kieu id-tu-tang khong cho phep.

    // ---- 5. CON BUG: value object kha bien bi chia se ----
    // O C++ phai CO Y moi tao duoc bug nay (can shared_ptr hoac tham chieu), vi mac
    // dinh cua ngon ngu la sao chep. Do la loi the that su cua C++ o bai nay.
    auto chung = std::make_shared<DiaChiSai>(DiaChiSai{"12 Le Loi"});
    auto don1 = chung, don2 = chung;             // vo tinh dung chung mot object
    don1->setDuong("45 Nguyen Hue");             // chi dinh sua don 1
    check(don2->duong == "45 Nguyen Hue", "don 2 bi doi dia chi theo — du khong ai dung vao no");
    // Hang cua don 2 vua duoc giao sai dia chi. Bug ALIASING, va no chi ton tai duoc vi
    // value object co setter.

    DiaChi cuaDon1("12 Le Loi", "Ben Nghe", "TP.HCM");
    DiaChi cuaDon2 = cuaDon1;                    // sao chep gia tri — an toan tuyet doi
    cuaDon1 = cuaDon1.voiDuong("45 Nguyen Hue");
    check(cuaDon1.duong() == "45 Nguyen Hue", "don 1 doi");
    check(cuaDon2.duong() == "12 Le Loi", "don 2 khong he han gi");

    // ---- 6. Value object mang LUAT, khong chi mang du lieu ----
    check(a.cong(b) == Tien(100000, "VND"), "cong cung te thi duoc");
    bool chan = false;
    try { a.cong(Tien(10, "USD")); } catch (const std::invalid_argument&) { chan = true; }
    check(chan, "cong khac te bi chan — luat nam TRONG kieu du lieu");
    // Bai 90 di sau vao rieng Money.

    // ---- 7. Cam bay rieng cua C++: `const` tren field ----
    // De ep bat bien, phan xa dau tien la khai bao `const long soTien_;`. Nhung lam vay
    // se XOA luon toan tu gan mac dinh, va `std::vector<Tien>` mat kha nang sort/erase:
    //     error: use of deleted function 'Tien& Tien::operator=(const Tien&)'
    // Do la ly do lop Tien o tren KHONG dung `const` tren field. Bat bien duoc bao dam
    // bang viec khong co setter va moi phuong thuc deu `const` — du an toan, va khong
    // hy sinh kha nang dung trong container.
    std::vector<Tien> dsTien{Tien(20000, "VND"), Tien(50000, "VND")};
    dsTien[0] = Tien(30000, "VND");   // gan duoc, vi khong co field `const`
    check(dsTien[0].soTien() == 30000, "value object nen gan lai duoc — dung `const` o cho khac");
    // Nguoc lai, `DiemGiao::ma_` CO `const`, va dung: dinh danh entity khong bao gio
    // duoc phep doi, ke ca qua toan tu gan.

    // ---- 8. Bang quyet dinh ----
    //
    //   Cau hoi                                     | Tra loi CO -> loai nao
    //   --------------------------------------------|-----------------------
    //   Doi het thuoc tinh, con la cung mot thu?     | ENTITY
    //   Hai cai giong het thi thay cho nhau duoc?    | VALUE OBJECT
    //   Nghiep vu can lich su cua CAI NAY?           | ENTITY
    //   Co the chia se tu do giua nhieu chu so huu?  | VALUE OBJECT
    //
    // Quy tac thuc dung: MAC DINH la value object. Chi nang len entity khi co mot cau
    // hoi nghiep vu that su can theo doi cai cu the do qua thoi gian.
    //
    // Va o C++, quy tac do co mot phien ban co the go ra thanh code:
    //   value object -> sao chep duoc, luu theo gia tri
    //   entity       -> `= delete` copy, luu qua unique_ptr
    check(true, "mac dinh la value object; entity phai co ly do");

    std::cout << "OK\n";
    return 0;
}

File: 82-entity-value-object/entity_value_demo.py
# Ngôn ngữ: Python
# Công dụng: Bản Python của cùng bài học — hai phép thử để quyết định entity hay
# value object, và các con bug thật mà chọn sai loại sinh ra: hai khách hàng chưa
# lưu gộp làm một, địa chỉ khả biến bị chia sẻ, entity bị nhân bản bởi một hàm
# tưởng là vô hại.
# Tại sao cần học: Python cho value object gần như miễn phí bằng
# `@dataclass(frozen=True)` — nhưng chính sự tiện lợi đó dựng ra cái bẫy lớn nhất:
# `@dataclass` MẶC ĐỊNH sinh `__eq__` so sánh TẤT CẢ field và đặt `__hash__ = None`.
# Dán nó lên một entity là bạn vừa nhận một lớp so sánh theo trạng thái khả biến và
# không bỏ vào `set` được. Bài cũng chỉ ra `dataclasses.replace()`: đúng tuyệt đối
# với value object, và là máy sinh bug khi gọi trên entity.

import copy
from dataclasses import dataclass, field, replace
from typing import List


# =====================================================================
# VALUE OBJECT — `frozen=True` cho đủ bốn thứ cần thiết trong một dòng
# =====================================================================
@dataclass(frozen=True)
class Tien:
    so_tien: int
    tien_te: str

    def __post_init__(self):
        if self.so_tien < 0:
            raise ValueError("số tiền không được âm")
        if len(self.tien_te) != 3:
            raise ValueError("mã tiền tệ phải đúng 3 ký tự")

    # Value object mang LUẬT của chính nó, không chỉ mang dữ liệu.
    def cong(self, khac):
        if self.tien_te != khac.tien_te:
            raise ValueError("không cộng được hai loại tiền tệ")
        return Tien(self.so_tien + khac.so_tien, self.tien_te)   # TRẢ VỀ CÁI MỚI


@dataclass(frozen=True)
class DiaChi:
    duong: str
    phuong: str
    tinh: str

    def __post_init__(self):
        if not self.duong.strip():
            raise ValueError("đường không được rỗng")

    # "Đổi" một value object = tạo cái mới. Ở Python không cần viết `voi_duong()`
    # bằng tay — `replace()` làm sẵn (xem phần 6).


# =====================================================================
# ENTITY — `eq=False` để KHÔNG lấy equals của dataclass, rồi tự viết theo định danh
# =====================================================================
@dataclass(eq=False)
class DiemGiao:
    ma: str            # ĐỊNH DANH: gán lúc tạo, không bao giờ đổi
    dia_chi: DiaChi    # thuộc tính: đổi thoải mái
    nguoi_phu_trach: str

    def __post_init__(self):
        if not self.ma.strip():
            raise ValueError("điểm giao phải có mã ngay lúc tạo")

    # Entity so sánh CHỈ theo định danh. Không field nào khác được xuất hiện ở đây —
    # nếu có, sửa một thuộc tính là mất phần tử trong `set` (bài 75).
    def __eq__(self, khac):
        return isinstance(khac, DiemGiao) and self.ma == khac.ma

    def __hash__(self):
        return hash(self.ma)


# =====================================================================
# BẢN SAI 1 — dán `@dataclass` mặc định lên một entity
# =====================================================================
@dataclass
class DiemGiaoSai:
    ma: str
    dia_chi: str
    nguoi_phu_trach: str
    # Không viết dòng nào, nhưng `@dataclass` vừa lặng lẽ làm hai việc:
    #   1. sinh `__eq__` so sánh CẢ BA field — kể cả hai field khả biến;
    #   2. đặt `__hash__ = None` -> lớp này KHÔNG bỏ vào `set`/`dict` được.


# =====================================================================
# BẢN SAI 2 — entity lấy định danh từ CSDL
# =====================================================================
@dataclass
class KhachHangSai:
    ten: str
    id: int = 0        # 0 = "chưa lưu". Quy ước phổ biến nhất, và sai nhất.

    def __eq__(self, k):
        return isinstance(k, KhachHangSai) and self.id == k.id

    def __hash__(self):
        return hash(self.id)


# BẢN ĐÚNG — định danh sinh trong miền, có ngay từ lúc tạo
class KhachHang:
    _dem = 0

    def __init__(self, ten):
        KhachHang._dem += 1
        self.ma = f"KH-{KhachHang._dem}"    # có định danh TRƯỚC khi chạm CSDL
        self.ten = ten

    def __eq__(self, k):
        return isinstance(k, KhachHang) and self.ma == k.ma

    def __hash__(self):
        return hash(self.ma)


# =====================================================================
# BẢN SAI 3 — "value object" nhưng khả biến
# =====================================================================
@dataclass
class DiaChiSai:
    duong: str


@dataclass
class DonHangSai:
    dia_chi_giao: DiaChiSai


# ---- Self-check ----
if __name__ == "__main__":
    # ---- 1. HAI PHÉP THỬ để quyết định entity hay value object ----
    #
    # Phép thử A — "đổi HẾT thuộc tính, còn là cùng một thứ không?"
    kho = DiemGiao("DG-01", DiaChi("12 Lê Lợi", "Bến Nghé", "TP.HCM"), "anh Nam")
    tham_chieu_cu = kho
    kho.dia_chi = DiaChi("45 Nguyễn Huệ", "Bến Nghé", "TP.HCM")
    kho.nguoi_phu_trach = "chị Lan"
    assert kho == tham_chieu_cu, "đổi hết thuộc tính, vẫn là cùng điểm giao -> ENTITY"

    # Phép thử B — "hai cái giống hệt nhau, thay cho nhau được không?"
    a, b = Tien(50_000, "VND"), Tien(50_000, "VND")
    assert a == b, "hai tờ 50.000đ thay cho nhau được -> VALUE OBJECT"
    assert a is not b, "vẫn là hai object khác nhau — và điều đó KHÔNG quan trọng"

    # ---- 2. CÙNG MỘT KHÁI NIỆM, HAI VAI TRÒ — tuỳ NGỮ CẢNH ----
    dc_don = DiaChi("12 Lê Lợi", "Bến Nghé", "TP.HCM")
    dc_khac = DiaChi("12 Lê Lợi", "Bến Nghé", "TP.HCM")
    assert dc_don == dc_khac, "trong đơn hàng: hai địa chỉ giống nhau LÀ MỘT"

    kho_a = DiemGiao("DG-01", dc_don, "anh Nam")
    kho_b = DiemGiao("DG-02", dc_khac, "chị Lan")
    assert kho_a != kho_b, "trong vận chuyển: cùng địa chỉ vẫn là HAI điểm giao"
    assert kho_a.dia_chi == kho_b.dia_chi, "dù thuộc tính địa chỉ của chúng bằng nhau"
    # Câu hỏi "cái này là entity hay value object" KHÔNG có câu trả lời chung. Nó phụ
    # thuộc vào việc nghiệp vụ có cần theo dõi CÁI CỤ THỂ NÀY qua thời gian hay không.

    # ---- 3. CÁI BẪY LỚN NHẤT CỦA PYTHON: `@dataclass` mặc định trên entity ----
    s1 = DiemGiaoSai("DG-01", "12 Lê Lợi", "anh Nam")
    s2 = DiemGiaoSai("DG-01", "12 Lê Lợi", "anh Nam")
    assert s1 == s2, "hai object này bằng nhau..."
    s2.nguoi_phu_trach = "chị Lan"
    assert s1 != s2, "...cho tới khi đổi người phụ trách. CÙNG MÃ mà không còn bằng nhau"
    # Đọc lại: hai bản ghi của CÙNG MỘT điểm giao (mã DG-01) vừa trở thành hai thứ khác
    # nhau, chỉ vì một thuộc tính đổi. Đó là định nghĩa của việc mất định danh.

    khong_bam_duoc = False
    try:
        hash(s1)
    except TypeError:
        khong_bam_duoc = True
    assert khong_bam_duoc, "`@dataclass` mặc định đặt __hash__ = None -> không bỏ vào set được"
    # Hai dòng trên là hệ quả của MỘT dòng `@dataclass` không tham số. Python không cảnh
    # báo gì; lỗi chỉ nổ khi ai đó thử `set(cac_diem_giao)` — thường là ở production.
    #
    # Ba cách khai báo, ba ý nghĩa hoàn toàn khác nhau:
    #   @dataclass                  -> so theo mọi field, KHÔNG hash được   (túi dữ liệu)
    #   @dataclass(frozen=True)     -> so theo mọi field, hash được         (VALUE OBJECT)
    #   @dataclass(eq=False) + tự viết __eq__/__hash__ theo id  -> (ENTITY)

    # ---- 4. CON BUG: định danh do CSDL cấp -> hai khách gộp làm một ----
    gio_sai = {KhachHangSai("Nguyễn Văn A"), KhachHangSai("Trần Thị B")}
    assert len(gio_sai) == 1, "HAI khách hàng khác nhau, set chỉ giữ MỘT"
    # Mất một khách hàng. Không ngoại lệ, không log. Bug này chỉ xuất hiện khi xử lý
    # theo lô (nhập file CSV, tạo hàng loạt) — nghĩa là nó qua được mọi test thủ công.

    gio_dung = {KhachHang("Nguyễn Văn A"), KhachHang("Trần Thị B")}
    assert len(gio_dung) == 2, "định danh sinh trong miền -> hai khách, giữ đúng hai"
    # Hệ quả thực tế: test không cần CSDL, và có thể gửi entity qua hàng đợi trước khi
    # lưu — thứ mà kiểu id-tự-tăng không cho phép.

    # ---- 5. CON BUG: value object khả biến bị chia sẻ ----
    chung = DiaChiSai("12 Lê Lợi")
    don1, don2 = DonHangSai(chung), DonHangSai(chung)   # vô tình dùng chung một object
    don1.dia_chi_giao.duong = "45 Nguyễn Huệ"           # chỉ định sửa đơn 1
    assert don2.dia_chi_giao.duong == "45 Nguyễn Huệ", \
        "đơn 2 bị đổi địa chỉ theo — dù không ai đụng vào nó"
    # Hàng của đơn 2 vừa được giao sai địa chỉ. Bug ALIASING, và ở Python nó DỄ xảy ra
    # nhất trong ba ngôn ngữ, vì gán luôn là chia sẻ tham chiếu, không bao giờ là sao chép.

    cua_don1 = cua_don2 = DiaChi("12 Lê Lợi", "Bến Nghé", "TP.HCM")
    cua_don1 = replace(cua_don1, duong="45 Nguyễn Huệ")   # tạo CÁI MỚI, gán lại
    assert cua_don1.duong == "45 Nguyễn Huệ", "đơn 1 đổi"
    assert cua_don2.duong == "12 Lê Lợi", "đơn 2 không hề hấn gì"
    # Vì bất biến, chia sẻ object là AN TOÀN — thậm chí còn tiết kiệm bộ nhớ.

    # ---- 6. `dataclasses.replace()`: đúng với VO, là máy sinh bug với ENTITY ----
    t2 = replace(a, so_tien=70_000)
    assert t2 == Tien(70_000, "VND") and a == Tien(50_000, "VND"), \
        "replace trên value object: đúng hoàn toàn, bản gốc không đổi"

    kho_nhan_ban = replace(kho_a, nguoi_phu_trach="chị Lan")
    assert kho_nhan_ban == kho_a, "hai object này LÀ CÙNG MỘT điểm giao (cùng mã)..."
    assert kho_nhan_ban.nguoi_phu_trach != kho_a.nguoi_phu_trach, "...nhưng trạng thái KHÁC NHAU"
    assert kho_nhan_ban is not kho_a, "và là hai object rời nhau trong bộ nhớ"
    # Cùng một cái kho, đồng thời do anh Nam và chị Lan phụ trách. Ngoài đời không tồn
    # tại. Từ đây trở đi, mọi thay đổi trên bản này sẽ không đến bản kia — dữ liệu trôi
    # ra khỏi nhau, và không ai biết bản nào mới đúng.
    #
    # `copy.copy()` và `copy.deepcopy()` gây đúng cùng một chuyện:
    ban_sao = copy.deepcopy(kho_a)
    assert ban_sao == kho_a and ban_sao is not kho_a, "deepcopy cũng nhân bản định danh"
    #
    # C++ dập được lỗi này ở mức trình biên dịch (`DiemGiao(const DiemGiao&) = delete`).
    # Python thì không có công cụ tương đương — nên nó phải thành LUẬT ĐỘI: entity chỉ
    # được sửa tại chỗ, không bao giờ đi qua `replace`/`copy`. Nếu muốn ép bằng máy, cách
    # thực dụng nhất là chặn ngay trong lớp:
    #     def __copy__(self): raise RuntimeError("không được nhân bản entity")
    #     def __deepcopy__(self, memo): raise RuntimeError("không được nhân bản entity")

    # ---- 7. Cạm bẫy: `frozen=True` là BẤT BIẾN NÔNG ----
    @dataclass(frozen=True)
    class GioHang:
        ma: str
        mat_hang: List[str] = field(default_factory=list)

    ds = ["bút"]
    gh = GioHang("GH-1", ds)
    ds.append("vở")                       # sửa từ BÊN NGOÀI dataclass
    assert len(gh.mat_hang) == 2, "frozen=True KHÔNG làm cho list bên trong bất biến"
    # `frozen` chỉ chặn việc GÁN LẠI field, không chặn việc sửa thứ field đó trỏ tới
    # (bài 73). Và hệ quả kéo theo: object này hash không được, dù mang tiếng frozen.
    khong_bam_duoc = False
    try:
        hash(gh)
    except TypeError:
        khong_bam_duoc = True
    assert khong_bam_duoc, "frozen chứa list -> hash() vẫn ném TypeError"

    @dataclass(frozen=True)
    class GioHangDung:
        ma: str
        mat_hang: tuple = ()

        def __post_init__(self):
            object.__setattr__(self, "mat_hang", tuple(self.mat_hang))   # sao chép phòng vệ

    ds2 = ["bút"]
    ghd = GioHangDung("GH-2", ds2)
    ds2.append("vở")
    assert len(ghd.mat_hang) == 1, "chuyển sang tuple chặn được rò rỉ khả biến"
    assert hash(ghd) == hash(GioHangDung("GH-2", ["bút"])), "và hash lại dùng được"
    # Quy tắc: value object trong Python chỉ được chứa `tuple`/`frozenset`/value object
    # khác. Thấy `List` hay `Dict` trong một dataclass frozen là thấy một cái bẫy.

    # ---- 8. Bảng quyết định ----
    #
    #   Câu hỏi                                         | Trả lời CÓ -> loại nào
    #   ------------------------------------------------|-----------------------
    #   Đổi hết thuộc tính, còn là cùng một thứ?         | ENTITY
    #   Hai cái giống hệt thì thay cho nhau được?        | VALUE OBJECT
    #   Nghiệp vụ cần lịch sử của CÁI NÀY?               | ENTITY
    #   Có thể chia sẻ tự do giữa nhiều chủ sở hữu?      | VALUE OBJECT
    #
    # Quy tắc thực dụng: MẶC ĐỊNH là value object. Chỉ nâng lên entity khi có một câu
    # hỏi nghiệp vụ thật sự cần theo dõi cái cụ thể đó qua thời gian. Entity đắt hơn
    # nhiều — nó cần định danh, cần kho lưu trữ, cần vòng đời, và không chia sẻ tự do được.
    assert True, "mặc định là value object; entity phải có lý do"

    print("OK")
